machine: implement varlink interfaces io.systemd.MachineImage.{List, Update}

ikruglov opened the pull request:
This PR implements the following methods in varlink interface of machined:

- io.systemd.MachineImage.List
  it's an alternative to the following dbus methods:
    - GetImage
    - ListImages
    - GetImageHostname
    - GetImageMachineID
    - GetImageMachineInfo
    - GetImageOSRelease

- io.systemd.MachineImage.Update
  it's an alternative to the following dbus methods:
    - RenameImage
    - MarkImageReadOnly
    - SetImageLimit

ikruglov commented:
@DaanDeMeyer I've re-based this PR on top of yesterday's merged changes. Please have a look.

DaanDeMeyer commented on src/machine/machined-varlink.c:
```diff
@@ -507,6 +508,112 @@ static int vl_method_terminate(sd_varlink *link, sd_json_variant *parameters, sd
+
+        r = sd_json_buildo(
+                        &v,
+                        SD_JSON_BUILD_PAIR("name", SD_JSON_BUILD_STRING(image->name)),
```
We have `SD_JSON_BUILD_PAIR_STRING` as a shortcut for this

DaanDeMeyer commented on src/machine/machined-varlink.c:
```diff
@@ -507,6 +508,112 @@ static int vl_method_terminate(sd_varlink *link, sd_json_variant *parameters, sd
+                        SD_JSON_BUILD_PAIR("type", SD_JSON_BUILD_STRING(image_type_to_string(image->type))),
+                        SD_JSON_BUILD_PAIR("class", SD_JSON_BUILD_STRING(image_class_to_string(image->class))),
+                        SD_JSON_BUILD_PAIR("readOnly", SD_JSON_BUILD_BOOLEAN(image->read_only)),
+                        SD_JSON_BUILD_PAIR("creationTimestamp", SD_JSON_BUILD_UNSIGNED(image->crtime)),
```
`SD_JSON_BUILD_PAIR_UNSIGNED`

DaanDeMeyer commented on src/machine/machined-varlink.c:
```diff
@@ -681,7 +682,8 @@ static int manager_varlink_init_machine(Manager *m) {
                         "io.systemd.Machine.Terminate",  vl_method_terminate,
                         "io.systemd.Machine.Kill",       vl_method_kill,
-                        "io.systemd.Machine.ListImages", vl_method_list_images);
+                        "io.systemd.Machine.ListImages", vl_method_list_images,
+                        "io.systemd.Machine.UpdateImage", vl_method_update_image);
```
Please keep the alignment intact here

DaanDeMeyer commented on src/machine/image-varlink.h:
```diff
@@ -0,0 +1,8 @@
+
+#include "sd-varlink.h"
+
+#include "machined.h"
```
This include seems unnecessary?

DaanDeMeyer commented on src/machine/image-varlink.c:
```diff
@@ -0,0 +1,105 @@
+        }
+
+        if (p.read_only >= 0) {
+                r = image_read_only(image, (bool) p.read_only);
```
Is the explicit cast required here?

DaanDeMeyer commented on src/machine/image-varlink.c:
```diff
@@ -0,0 +1,105 @@
+#include "image-varlink.h"
+#include "string-util.h"
+
+static int rename_image(Manager *manager, Image *image, const char* new_name) {
```
Please reuse this for the dbus method as well

ikruglov commented:
I'd like to highlight two questions:

1. [When to load metadata in ListImages()](https://github.com/systemd/systemd/pull/34623/files#diff-15c4a737d22304facac2b3af8fc398e63e92c1e3ed63b4bc975a853cf8c7ab17R580).
I currently implemented it in a way where image metadata is loaded on a point query `ListImages '{"name": "foo"}'`, but not in a range query `ListImages '{}'`. My reasoning is that loading metadata is potentially a heavy operation (disk IO) on a loaded machine. Alternatively, I can also imagine having an extra flag: `ListImages '{ "includeMetadata": true }'` to explicitly request loading metadata for all images.

2. [New polkit verb "update_image"](https://github.com/systemd/systemd/pull/34623/files#diff-ae2c8b6e62bf49ab433c6c5c42ff15b987fb35cfee045993e355455ae3355e62R60)
The PR introduced new polkit verb "update_image". I couldn't reuse dbus one because there are three separate verbs: "mark_read_only", "rename", "set_limit".

Looking for your feedback for both of these questions.

poettering commented on src/machine/machined-core.c:
```diff
@@ -456,3 +456,22 @@ int manager_acquire_image(Manager *m, const char *name, Image **ret) {
+        /* The image is cached with its name, hence it is necessary to remove from the cache before renaming. */
+        assert_se(hashmap_remove_value(m->image_cache, image->name, image));
+
+        int r = image_rename(image, new_name);
```
as a special rule we always declare `int r;` at outermost function scope. `r` is special to us, it's where we store error codes we receive to, always, hence this special treatment. hence please do `int r;` at the top of the funcion

ikruglov replied:
np

poettering commented on src/machine/machined-core.c:
```diff
@@ -456,3 +456,22 @@ int manager_acquire_image(Manager *m, const char *name, Image **ret) {
+        }
+
+        /* Then save the object again in the cache. */
+        assert_se(hashmap_put(m->image_cache, image->name, image) > 0);
```
this can fail because it involves a memory allocation. I understand this is a pre-existing bug, but if you touch this please fix this. given the cache is a "best effort" thing anyway, maybe on any kind of error just debug or warn log, free the cache, and release the image object and continue.

ikruglov replied:
Is this what you're thinking about?

```
          /* Then save the object again in the cache. */
          r = hashmap_put(m->image_cache, image->name, image);
          if (r < 0) {
                  image_unref(image);
                  return log_debug_errno(r, "Failed to put image into cache: %m");
          }
```

poettering replied:
and i think we should do `image = image_unref(image)`, just as a matter of hygiene. even if we never use `image` afterwards, it's a lot cleaner if we invalidate pointers that we dropped our ref of.

poettering commented on src/machine/machined-varlink.c:
```diff
@@ -507,6 +508,112 @@ static int vl_method_terminate(sd_varlink *link, sd_json_variant *parameters, sd
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to find image: %m");
+
+                // TODO(ikruglov): decide if we want to load metadata in the general case of listing
```
I'd really make this an option. i.e. add a boolean argument called acquireMetadata or so, which defaults to false

poettering replied:
not sure i grok your question? just make it a boolean field passed to the call which we apply to all machines we enumerate

poettering commented on src/shared/varlink-io.systemd.Machine.c:
```diff
@@ -93,6 +130,10 @@ SD_VARLINK_DEFINE_INTERFACE(
                 SD_VARLINK_SYMBOL_COMMENT("List running machines"),
                 &vl_method_List,
+                SD_VARLINK_SYMBOL_COMMENT("List images"),
+                &vl_method_ListImages,
```
I'd prefer if we'd add a second interface for covering images. i.e. add io.systemd.MachineImage or so for this, so that the methods can then be called "List" instead of "ListImages" too.

(Sooner or later we want to cover unpriv containers nicely with machined, i.e. images i drop somewhere in $HOME. That probably means we want two instances of machined: a system wide and a per-user once. But otoh we always want to register containers on the system so that we can support hostname resolution for container names nicely, that creates this weird split that we want images per user but live machine registration for the system. Hence, as preparation for that I think it would be much nicer if we split this all up in two interfaces already to be ready for this.)

note that varlink services can implemented as many interfaces as they want on the same socket, that's entirely fine.

poettering commented on src/machine/image-varlink.c:
```diff
@@ -0,0 +1,84 @@
+        };
+
+        static const sd_json_dispatch_field dispatch_table[] = {
+                { "name",     SD_JSON_VARIANT_STRING,        sd_json_dispatch_const_string, offsetof(struct params, image_name), SD_JSON_MANDATORY },
```
the other calls also allowed referencing the machine by pid. I think just to be systematic we should retain that here, even if that doesn't make too much sense to actually IRL.

ikruglov replied:
here I'm referencing image, not machine. Does it make sense to reference by pid?

poettering replied:
oh, right, ignore me. i am an idiot

poettering commented on src/machine/image-varlink.c:
```diff
@@ -0,0 +1,84 @@
+                r = image_set_limit(image, p.limit);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to set image limit: %m");
+        }
```
I think if the latter two fail we should still execute the other. right now, if the r/o thing fails you won't attempt to apply the limit, but i think we should.

ikruglov replied:
how do you imagine error handing in this case? Just log_debug() or accumulate all errors and return them all at the end?

poettering replied:
i think the rename thing should always be fatal, immediately. the other two operations we should both execute and then report the "first" error we encounter. Which one you make first is up to you i guess.

poettering commented on src/shared/varlink-io.systemd.Machine.c:
```diff
@@ -113,6 +113,17 @@ static SD_VARLINK_DEFINE_METHOD_FULL(
+                SD_VARLINK_FIELD_COMMENT("If non-null value of the read-only flag of the image"),
+                SD_VARLINK_DEFINE_INPUT(readOnly, SD_VARLINK_BOOL, SD_VARLINK_NULLABLE),
+                SD_VARLINK_FIELD_COMMENT("If non-null value of image quota limit"),
+                SD_VARLINK_DEFINE_INPUT(limit, SD_VARLINK_INT, SD_VARLINK_NULLABLE));
```
the interactiveAuthAllowed field is missing

poettering commented:
(btw, i hope you saw my post-merge comments on that other PR? please address those too)

ikruglov commented:
@poettering, I've addressed all your suggestions. Please have a look.

ikruglov commented:
> (btw, i hope you saw my post-merge comments on that other PR? please address those too)

I've started addressing your suggestions in https://github.com/systemd/systemd/pull/34681. Need your input to complete it though.

ikruglov commented:
@poettering @DaanDeMeyer I'd appreciate another look for this PR.

poettering commented on src/machine/machined-varlink.c:
```diff
@@ -507,6 +508,126 @@ static int vl_method_terminate(sd_varlink *link, sd_json_variant *parameters, sd
+                        SD_JSON_BUILD_PAIR("class", SD_JSON_BUILD_STRING(image_class_to_string(image->class))),
+                        SD_JSON_BUILD_PAIR_BOOLEAN("readOnly", image->read_only),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("creationTimestamp", image->crtime),
+                        JSON_BUILD_PAIR_UNSIGNED_NON_ZERO("modificationTimestamp", image->mtime),
```
*both* the mtime and the *crtime* are not supported by all image backends. hence should both be using JSON_BUILD_PAIR_UNSIGNED_NON_ZERO()

poettering commented on src/machine/machined-varlink.c:
```diff
@@ -507,6 +508,126 @@ static int vl_method_terminate(sd_varlink *link, sd_json_variant *parameters, sd
+                        SD_JSON_BUILD_PAIR_STRING("name", image->name),
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("path", image->path),
+                        SD_JSON_BUILD_PAIR("type", SD_JSON_BUILD_STRING(image_type_to_string(image->type))),
+                        SD_JSON_BUILD_PAIR("class", SD_JSON_BUILD_STRING(image_class_to_string(image->class))),
```
you could use SD_JSON_BUILD_PAIR_STRING() here to shorten the line, but doesn't really matter

poettering commented on src/machine/machined-varlink.c:
```diff
@@ -507,6 +508,126 @@ static int vl_method_terminate(sd_varlink *link, sd_json_variant *parameters, sd
+                                                     JSON_BUILD_STRV_ENV_PAIR(image->machine_info)),
+                        SD_JSON_BUILD_PAIR_CONDITION(image->metadata_valid && image->os_release,
+                                                     "OSRelease",
+                                                     JSON_BUILD_STRV_ENV_PAIR(image->os_release)));
```
doesn't matter, but instead of repeating the image->metadata_valid check here each time, you could also do that in a separate sd_json_variant_merge_objectbo() call.

i.e.:

```c
if (image->metadata_valid) {
        r = sd_json_variant_merge_objectbo(&v, SD_JSON_BUILD_PAIR_CONDITION(…));
        if (r < 0)
                return r;
}
```

but doesn't really matter. Might be a bit easier to read though

poettering commented on src/machine/machined-varlink.c:
```diff
@@ -507,6 +508,126 @@ static int vl_method_terminate(sd_varlink *link, sd_json_variant *parameters, sd
+        return list_image_one(link, image, more);
+}
+
+
```
spurious double empty line

ikruglov commented:
@poettering I've implemented all your minor suggestions. Your comment about double empty line made me realize that function `list_image_one()` is redundant. So, I squashed `list_image_one()` and `list_image_one_and_maybe_read_metadata()` together.

poettering commented on src/machine/machined-varlink.c:
```diff
@@ -688,7 +689,8 @@ static int manager_varlink_init_machine(Manager *m) {
                         "io.systemd.Machine.Kill",        vl_method_kill,
-                        "io.systemd.MachineImage.List",   vl_method_list_images);
+                        "io.systemd.MachineImage.List",   vl_method_list_images,
+                        "io.systemd.MachineImage.Update", vl_method_update_image);
```
hmm, could you also add `r = sd_varlink_server_listen_address(s, "/run/systemd/machine/io.systemd.MachineImage", 0666);` below, so that we have a different entrypoint for the 2nd interface (even if we server everything on both).

(And then update the test case to use the 2nd entrypoint, just to be fully correct)

poettering commented:
sorry, found one more thing

ikruglov commented:
> sorry, found one more thing

done

yuwata commented on src/machine/machined-varlink.c:
```diff
@@ -508,6 +509,118 @@ static int vl_method_terminate(sd_varlink *link, sd_json_variant *parameters, sd
+                                JSON_BUILD_PAIR_STRING_NON_EMPTY("hostname", image->hostname),
+                                SD_JSON_BUILD_PAIR_CONDITION(!sd_id128_is_null(image->machine_id), "machineId", SD_JSON_BUILD_ID128(image->machine_id)),
+                                SD_JSON_BUILD_PAIR_CONDITION(!!image->machine_info, "machineInfo", JSON_BUILD_STRV_ENV_PAIR(image->machine_info)),
+                                SD_JSON_BUILD_PAIR_CONDITION(!!image->os_release, "OSRelease", JSON_BUILD_STRV_ENV_PAIR(image->os_release)));
```
`!strv_isempty()` is better?

yuwata commented on src/machine/machined-varlink.c:
```diff
@@ -508,6 +509,118 @@ static int vl_method_terminate(sd_varlink *link, sd_json_variant *parameters, sd
+        Image *image;
+        int r;
+
+        assert(parameters);
```
assert(link);

yuwata commented on src/machine/image-varlink.c:
```diff
@@ -0,0 +1,88 @@
+                        log_debug_errno(r, "Failed to set image limit, ignoring: %m");
+        }
+
+        /* We intentionally swallowed errors from image_read_only() and image_set_limit(). Here we return last one to the user if any */
+        if (r < 0)
+                return r;
```
Then, maybe better to use RET_GATHER() in the above, and return the first error?

yuwata commented on src/machine/machined-varlink.c:
```diff
@@ -508,6 +509,118 @@ static int vl_method_terminate(sd_varlink *link, sd_json_variant *parameters, sd
         return lookup_machine_and_call_method(link, parameters, flags, userdata, vl_method_terminate_internal);
 }
 
+static int list_image_one_and_maybe_read_metadata(sd_varlink *link, Image *image, bool more, bool read_metadata) {
```
I'd like to move this and below to image-varlink.c

ikruglov replied:
I don't have a strong opinion, but the same DBus logic is in machined-dbus.c. So, I copied it here to be consistent. I understand that DBus has per-object methods, which make a situation a bit different. Yet, I'd keep it here. Or we would also need to move `list_machine_one_and_maybe_read_metadata()` to machine-varlink.c for consistency. If so, I better do it in a separate PR.

yuwata replied:
OK, feel free to keep them. We can move them later.

ikruglov replied:
I've realized that `list_machine_one_and_maybe_read_metadata()` is from upcoming PR. Just FYI.

yuwata reviewed (changes requested):
Looks mostly good. But several minor comments.

ikruglov commented:
@yuwata @poettering I've fixed all suggestions. Hope for a merge. Thanks

mrc0mmand commented:
Btw, CI was not really happy:

```
Oct 14 09:33:53 H systemd-machined[652]: ==652==ERROR: LeakSanitizer: detected memory leaks
Oct 14 09:33:53 H systemd-machined[652]: Direct leak of 960 byte(s) in 6 object(s) allocated from:
Oct 14 09:33:53 H systemd-machined[652]:     #0 0x78b2d4d58009 in malloc (/usr/lib/clang/18/lib/linux/libclang_rt.asan-x86_64.so+0x158009) (BuildId: 0f6d876ea3e9400890133e57c33f1b3cff646b22)
Oct 14 09:33:53 H systemd-machined[652]:     #1 0x78b2d3e8a531 in malloc_multiply /systemd-meson-build/../build/src/basic/alloc-util.h:119:16
Oct 14 09:33:53 H systemd-machined[652]:     #2 0x78b2d3e8a531 in image_new /systemd-meson-build/../build/src/shared/discover-image.c:196:13
Oct 14 09:33:53 H systemd-machined[652]:     #3 0x78b2d3e7e771 in image_make /systemd-meson-build/../build/src/shared/discover-image.c:429:21
Oct 14 09:33:53 H systemd-machined[652]:     #4 0x78b2d3e7d52b in image_find /systemd-meson-build/../build/src/shared/discover-image.c:708:29
Oct 14 09:33:53 H systemd-machined[652]:     #5 0x585b1a08bed9 in vl_method_list_images /systemd-meson-build/../build/src/machine/machined-varlink.c:589:21
Oct 14 09:33:53 H systemd-machined[652]:     #6 0x78b2d44da37b in varlink_dispatch_method /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1383:29
Oct 14 09:33:53 H systemd-machined[652]:     #7 0x78b2d44da37b in sd_varlink_process /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1456:13
Oct 14 09:33:53 H systemd-machined[652]:     #8 0x78b2d44f41f5 in defer_callback /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:2880:16
Oct 14 09:33:53 H systemd-machined[652]:     #9 0x78b2d4367bcf in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c
Oct 14 09:33:53 H systemd-machined[652]:     #10 0x78b2d43660d9 in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4849:21
Oct 14 09:33:53 H systemd-machined[652]:     #11 0x78b2d436a039 in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4910:21
Oct 14 09:33:53 H systemd-machined[652]:     #12 0x78b2d3e07515 in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:147:21
Oct 14 09:33:53 H systemd-machined[652]:     #13 0x585b1a06cc1e in run /systemd-meson-build/../build/src/machine/machined.c:346:13
Oct 14 09:33:53 H systemd-machined[652]:     #14 0x585b1a06cc1e in main /systemd-meson-build/../build/src/machine/machined.c:358:1
Oct 14 09:33:53 H systemd-machined[652]:     #15 0x78b2d3634e07  (/usr/lib/libc.so.6+0x25e07) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #16 0x78b2d3634ecb in __libc_start_main (/usr/lib/libc.so.6+0x25ecb) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #17 0x585b1a06a294 in _start (/usr/lib/systemd/systemd-machined+0x34294) (BuildId: 7e799bc6af64f913587709a1e143113c9443cdb3)
Oct 14 09:33:53 H systemd-machined[652]: Indirect leak of 512 byte(s) in 8 object(s) allocated from:
Oct 14 09:33:53 H systemd-machined[652]:     #0 0x78b2d4d5719a  (/usr/lib/clang/18/lib/linux/libclang_rt.asan-x86_64.so+0x15719a) (BuildId: 0f6d876ea3e9400890133e57c33f1b3cff646b22)
Oct 14 09:33:53 H systemd-machined[652]:     #1 0x78b2d40faec5 in greedy_realloc /systemd-meson-build/../build/src/basic/alloc-util.c:69:13
Oct 14 09:33:53 H systemd-machined[652]:     #2 0x78b2d4134444 in parse_env_file_internal /systemd-meson-build/../build/src/basic/env-file.c:208:38
Oct 14 09:33:53 H systemd-machined[652]:     #3 0x78b2d413643c in load_env_file_pairs /systemd-meson-build/../build/src/basic/env-file.c:500:13
Oct 14 09:33:53 H systemd-machined[652]:     #4 0x78b2d4136ab4 in load_env_file_pairs_fd /systemd-meson-build/../build/src/basic/env-file.c:518:16
Oct 14 09:33:53 H systemd-machined[652]:     #5 0x78b2d41b7383 in load_extension_release_pairs /systemd-meson-build/../build/src/basic/os-util.c:434:16
Oct 14 09:33:53 H systemd-machined[652]:     #6 0x78b2d3e881ab in load_os_release_pairs /systemd-meson-build/../build/src/basic/os-util.h:54:16
Oct 14 09:33:53 H systemd-machined[652]:     #7 0x78b2d3e881ab in image_read_metadata /systemd-meson-build/../build/src/shared/discover-image.c:1502:21
Oct 14 09:33:53 H systemd-machined[652]:     #8 0x585b1a08d236 in list_image_one_and_maybe_read_metadata /systemd-meson-build/../build/src/machine/machined-varlink.c:520:21
Oct 14 09:33:53 H systemd-machined[652]:     #9 0x585b1a08c082 in vl_method_list_images /systemd-meson-build/../build/src/machine/machined-varlink.c:595:24
Oct 14 09:33:53 H systemd-machined[652]:     #10 0x78b2d44da37b in varlink_dispatch_method /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1383:29
Oct 14 09:33:53 H systemd-machined[652]:     #11 0x78b2d44da37b in sd_varlink_process /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1456:13
Oct 14 09:33:53 H systemd-machined[652]:     #12 0x78b2d44f41f5 in defer_callback /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:2880:16
Oct 14 09:33:53 H systemd-machined[652]:     #13 0x78b2d4367bcf in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c
Oct 14 09:33:53 H systemd-machined[652]:     #14 0x78b2d43660d9 in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4849:21
Oct 14 09:33:53 H systemd-machined[652]:     #15 0x78b2d436a039 in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4910:21
Oct 14 09:33:53 H systemd-machined[652]:     #16 0x78b2d3e07515 in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:147:21
Oct 14 09:33:53 H systemd-machined[652]:     #17 0x585b1a06cc1e in run /systemd-meson-build/../build/src/machine/machined.c:346:13
Oct 14 09:33:53 H systemd-machined[652]:     #18 0x585b1a06cc1e in main /systemd-meson-build/../build/src/machine/machined.c:358:1
Oct 14 09:33:53 H systemd-machined[652]:     #19 0x78b2d3634e07  (/usr/lib/libc.so.6+0x25e07) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #20 0x78b2d3634ecb in __libc_start_main (/usr/lib/libc.so.6+0x25ecb) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #21 0x585b1a06a294 in _start (/usr/lib/systemd/systemd-machined+0x34294) (BuildId: 7e799bc6af64f913587709a1e143113c9443cdb3)
Oct 14 09:33:53 H systemd-machined[652]: Indirect leak of 256 byte(s) in 1 object(s) allocated from:
Oct 14 09:33:53 H systemd-machined[652]:     #0 0x78b2d4d572df in reallocarray (/usr/lib/clang/18/lib/linux/libclang_rt.asan-x86_64.so+0x1572df) (BuildId: 0f6d876ea3e9400890133e57c33f1b3cff646b22)
Oct 14 09:33:53 H systemd-machined[652]:     #1 0x78b2d4219e64 in strv_push_with_size /systemd-meson-build/../build/src/basic/strv.c:581:20
Oct 14 09:33:53 H systemd-machined[652]:     #2 0x78b2d41340d5 in parse_env_file_internal /systemd-meson-build/../build/src/basic/env-file.c:153:37
Oct 14 09:33:53 H systemd-machined[652]:     #3 0x78b2d413643c in load_env_file_pairs /systemd-meson-build/../build/src/basic/env-file.c:500:13
Oct 14 09:33:53 H systemd-machined[652]:     #4 0x78b2d4136ab4 in load_env_file_pairs_fd /systemd-meson-build/../build/src/basic/env-file.c:518:16
Oct 14 09:33:53 H systemd-machined[652]:     #5 0x78b2d41b7383 in load_extension_release_pairs /systemd-meson-build/../build/src/basic/os-util.c:434:16
Oct 14 09:33:53 H systemd-machined[652]:     #6 0x78b2d3e881ab in load_os_release_pairs /systemd-meson-build/../build/src/basic/os-util.h:54:16
Oct 14 09:33:53 H systemd-machined[652]:     #7 0x78b2d3e881ab in image_read_metadata /systemd-meson-build/../build/src/shared/discover-image.c:1502:21
Oct 14 09:33:53 H systemd-machined[652]:     #8 0x585b1a08d236 in list_image_one_and_maybe_read_metadata /systemd-meson-build/../build/src/machine/machined-varlink.c:520:21
Oct 14 09:33:53 H systemd-machined[652]:     #9 0x585b1a08c082 in vl_method_list_images /systemd-meson-build/../build/src/machine/machined-varlink.c:595:24
Oct 14 09:33:53 H systemd-machined[652]:     #10 0x78b2d44da37b in varlink_dispatch_method /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1383:29
Oct 14 09:33:53 H systemd-machined[652]:     #11 0x78b2d44da37b in sd_varlink_process /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1456:13
Oct 14 09:33:53 H systemd-machined[652]:     #12 0x78b2d44f41f5 in defer_callback /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:2880:16
Oct 14 09:33:53 H systemd-machined[652]:     #13 0x78b2d4367bcf in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c
Oct 14 09:33:53 H systemd-machined[652]:     #14 0x78b2d43660d9 in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4849:21
Oct 14 09:33:53 H systemd-machined[652]:     #15 0x78b2d436a039 in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4910:21
Oct 14 09:33:53 H systemd-machined[652]:     #16 0x78b2d3e07515 in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:147:21
Oct 14 09:33:53 H systemd-machined[652]:     #17 0x585b1a06cc1e in run /systemd-meson-build/../build/src/machine/machined.c:346:13
Oct 14 09:33:53 H systemd-machined[652]:     #18 0x585b1a06cc1e in main /systemd-meson-build/../build/src/machine/machined.c:358:1
Oct 14 09:33:53 H systemd-machined[652]:     #19 0x78b2d3634e07  (/usr/lib/libc.so.6+0x25e07) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #20 0x78b2d3634ecb in __libc_start_main (/usr/lib/libc.so.6+0x25ecb) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #21 0x585b1a06a294 in _start (/usr/lib/systemd/systemd-machined+0x34294) (BuildId: 7e799bc6af64f913587709a1e143113c9443cdb3)
Oct 14 09:33:53 H systemd-machined[652]: Indirect leak of 256 byte(s) in 6 object(s) allocated from:
Oct 14 09:33:53 H systemd-machined[652]:     #0 0x78b2d4d5719a  (/usr/lib/clang/18/lib/linux/libclang_rt.asan-x86_64.so+0x15719a) (BuildId: 0f6d876ea3e9400890133e57c33f1b3cff646b22)
Oct 14 09:33:53 H systemd-machined[652]:     #1 0x78b2d41be0dc in path_extend_internal /systemd-meson-build/../build/src/basic/path-util.c:585:14
Oct 14 09:33:53 H systemd-machined[652]:     #2 0x78b2d3e8a6ca in image_new /systemd-meson-build/../build/src/shared/discover-image.c:217:19
Oct 14 09:33:53 H systemd-machined[652]:     #3 0x78b2d3e7e771 in image_make /systemd-meson-build/../build/src/shared/discover-image.c:429:21
Oct 14 09:33:53 H systemd-machined[652]:     #4 0x78b2d3e7d52b in image_find /systemd-meson-build/../build/src/shared/discover-image.c:708:29
Oct 14 09:33:53 H systemd-machined[652]:     #5 0x585b1a08bed9 in vl_method_list_images /systemd-meson-build/../build/src/machine/machined-varlink.c:589:21
Oct 14 09:33:53 H systemd-machined[652]:     #6 0x78b2d44da37b in varlink_dispatch_method /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1383:29
Oct 14 09:33:53 H systemd-machined[652]:     #7 0x78b2d44da37b in sd_varlink_process /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1456:13
Oct 14 09:33:53 H systemd-machined[652]:     #8 0x78b2d44f41f5 in defer_callback /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:2880:16
Oct 14 09:33:53 H systemd-machined[652]:     #9 0x78b2d4367bcf in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c
Oct 14 09:33:53 H systemd-machined[652]:     #10 0x78b2d43660d9 in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4849:21
Oct 14 09:33:53 H systemd-machined[652]:     #11 0x78b2d436a039 in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4910:21
Oct 14 09:33:53 H systemd-machined[652]:     #12 0x78b2d3e07515 in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:147:21
Oct 14 09:33:53 H systemd-machined[652]:     #13 0x585b1a06cc1e in run /systemd-meson-build/../build/src/machine/machined.c:346:13
Oct 14 09:33:53 H systemd-machined[652]:     #14 0x585b1a06cc1e in main /systemd-meson-build/../build/src/machine/machined.c:358:1
Oct 14 09:33:53 H systemd-machined[652]:     #15 0x78b2d3634e07  (/usr/lib/libc.so.6+0x25e07) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #16 0x78b2d3634ecb in __libc_start_main (/usr/lib/libc.so.6+0x25ecb) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #17 0x585b1a06a294 in _start (/usr/lib/systemd/systemd-machined+0x34294) (BuildId: 7e799bc6af64f913587709a1e143113c9443cdb3)
Oct 14 09:33:53 H systemd-machined[652]: Indirect leak of 192 byte(s) in 3 object(s) allocated from:
Oct 14 09:33:53 H systemd-machined[652]:     #0 0x78b2d4d5719a  (/usr/lib/clang/18/lib/linux/libclang_rt.asan-x86_64.so+0x15719a) (BuildId: 0f6d876ea3e9400890133e57c33f1b3cff646b22)
Oct 14 09:33:53 H systemd-machined[652]:     #1 0x78b2d40faec5 in greedy_realloc /systemd-meson-build/../build/src/basic/alloc-util.c:69:13
Oct 14 09:33:53 H systemd-machined[652]:     #2 0x78b2d4134894 in parse_env_file_internal /systemd-meson-build/../build/src/basic/env-file.c:127:38
Oct 14 09:33:53 H systemd-machined[652]:     #3 0x78b2d413643c in load_env_file_pairs /systemd-meson-build/../build/src/basic/env-file.c:500:13
Oct 14 09:33:53 H systemd-machined[652]:     #4 0x78b2d4136ab4 in load_env_file_pairs_fd /systemd-meson-build/../build/src/basic/env-file.c:518:16
Oct 14 09:33:53 H systemd-machined[652]:     #5 0x78b2d41b7383 in load_extension_release_pairs /systemd-meson-build/../build/src/basic/os-util.c:434:16
Oct 14 09:33:53 H systemd-machined[652]:     #6 0x78b2d3e881ab in load_os_release_pairs /systemd-meson-build/../build/src/basic/os-util.h:54:16
Oct 14 09:33:53 H systemd-machined[652]:     #7 0x78b2d3e881ab in image_read_metadata /systemd-meson-build/../build/src/shared/discover-image.c:1502:21
Oct 14 09:33:53 H systemd-machined[652]:     #8 0x585b1a08d236 in list_image_one_and_maybe_read_metadata /systemd-meson-build/../build/src/machine/machined-varlink.c:520:21
Oct 14 09:33:53 H systemd-machined[652]:     #9 0x585b1a08c082 in vl_method_list_images /systemd-meson-build/../build/src/machine/machined-varlink.c:595:24
Oct 14 09:33:53 H systemd-machined[652]:     #10 0x78b2d44da37b in varlink_dispatch_method /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1383:29
Oct 14 09:33:53 H systemd-machined[652]:     #11 0x78b2d44da37b in sd_varlink_process /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1456:13
Oct 14 09:33:53 H systemd-machined[652]:     #12 0x78b2d44f41f5 in defer_callback /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:2880:16
Oct 14 09:33:53 H systemd-machined[652]:     #13 0x78b2d4367bcf in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c
Oct 14 09:33:53 H systemd-machined[652]:     #14 0x78b2d43660d9 in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4849:21
Oct 14 09:33:53 H systemd-machined[652]:     #15 0x78b2d436a039 in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4910:21
Oct 14 09:33:53 H systemd-machined[652]:     #16 0x78b2d3e07515 in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:147:21
Oct 14 09:33:53 H systemd-machined[652]:     #17 0x585b1a06cc1e in run /systemd-meson-build/../build/src/machine/machined.c:346:13
Oct 14 09:33:53 H systemd-machined[652]:     #18 0x585b1a06cc1e in main /systemd-meson-build/../build/src/machine/machined.c:358:1
Oct 14 09:33:53 H systemd-machined[652]:     #19 0x78b2d3634e07  (/usr/lib/libc.so.6+0x25e07) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #20 0x78b2d3634ecb in __libc_start_main (/usr/lib/libc.so.6+0x25ecb) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #21 0x585b1a06a294 in _start (/usr/lib/systemd/systemd-machined+0x34294) (BuildId: 7e799bc6af64f913587709a1e143113c9443cdb3)
Oct 14 09:33:53 H systemd-machined[652]: Indirect leak of 101 byte(s) in 8 object(s) allocated from:
Oct 14 09:33:53 H systemd-machined[652]:     #0 0x78b2d4d52682 in strdup (/usr/lib/clang/18/lib/linux/libclang_rt.asan-x86_64.so+0x152682) (BuildId: 0f6d876ea3e9400890133e57c33f1b3cff646b22)
Oct 14 09:33:53 H systemd-machined[652]:     #1 0x78b2d421ac7c in strv_extend_with_size /systemd-meson-build/../build/src/basic/strv.c:704:13
Oct 14 09:33:53 H systemd-machined[652]:     #2 0x78b2d4136802 in strv_extend /systemd-meson-build/../build/src/basic/strv.h:62:16
Oct 14 09:33:53 H systemd-machined[652]:     #3 0x78b2d4136802 in load_env_file_push_pairs /systemd-meson-build/../build/src/basic/env-file.c:483:13
Oct 14 09:33:53 H systemd-machined[652]:     #4 0x78b2d4133f1a in parse_env_file_internal /systemd-meson-build/../build/src/basic/env-file.c:110:37
Oct 14 09:33:53 H systemd-machined[652]:     #5 0x78b2d413643c in load_env_file_pairs /systemd-meson-build/../build/src/basic/env-file.c:500:13
Oct 14 09:33:53 H systemd-machined[652]:     #6 0x78b2d4136ab4 in load_env_file_pairs_fd /systemd-meson-build/../build/src/basic/env-file.c:518:16
Oct 14 09:33:53 H systemd-machined[652]:     #7 0x78b2d41b7383 in load_extension_release_pairs /systemd-meson-build/../build/src/basic/os-util.c:434:16
Oct 14 09:33:53 H systemd-machined[652]:     #8 0x78b2d3e881ab in load_os_release_pairs /systemd-meson-build/../build/src/basic/os-util.h:54:16
Oct 14 09:33:53 H systemd-machined[652]:     #9 0x78b2d3e881ab in image_read_metadata /systemd-meson-build/../build/src/shared/discover-image.c:1502:21
Oct 14 09:33:53 H systemd-machined[652]:     #10 0x585b1a08d236 in list_image_one_and_maybe_read_metadata /systemd-meson-build/../build/src/machine/machined-varlink.c:520:21
Oct 14 09:33:53 H systemd-machined[652]:     #11 0x585b1a08c082 in vl_method_list_images /systemd-meson-build/../build/src/machine/machined-varlink.c:595:24
Oct 14 09:33:53 H systemd-machined[652]:     #12 0x78b2d44da37b in varlink_dispatch_method /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1383:29
Oct 14 09:33:53 H systemd-machined[652]:     #13 0x78b2d44da37b in sd_varlink_process /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1456:13
Oct 14 09:33:53 H systemd-machined[652]:     #14 0x78b2d44f41f5 in defer_callback /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:2880:16
Oct 14 09:33:53 H systemd-machined[652]:     #15 0x78b2d4367bcf in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c
Oct 14 09:33:53 H systemd-machined[652]:     #16 0x78b2d43660d9 in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4849:21
Oct 14 09:33:53 H systemd-machined[652]:     #17 0x78b2d436a039 in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4910:21
Oct 14 09:33:53 H systemd-machined[652]:     #18 0x78b2d3e07515 in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:147:21
Oct 14 09:33:53 H systemd-machined[652]:     #19 0x585b1a06cc1e in run /systemd-meson-build/../build/src/machine/machined.c:346:13
Oct 14 09:33:53 H systemd-machined[652]:     #20 0x585b1a06cc1e in main /systemd-meson-build/../build/src/machine/machined.c:358:1
Oct 14 09:33:53 H systemd-machined[652]:     #21 0x78b2d3634e07  (/usr/lib/libc.so.6+0x25e07) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #22 0x78b2d3634ecb in __libc_start_main (/usr/lib/libc.so.6+0x25ecb) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #23 0x585b1a06a294 in _start (/usr/lib/systemd/systemd-machined+0x34294) (BuildId: 7e799bc6af64f913587709a1e143113c9443cdb3)
Oct 14 09:33:53 H systemd-machined[652]: Indirect leak of 94 byte(s) in 6 object(s) allocated from:
Oct 14 09:33:53 H systemd-machined[652]:     #0 0x78b2d4d52682 in strdup (/usr/lib/clang/18/lib/linux/libclang_rt.asan-x86_64.so+0x152682) (BuildId: 0f6d876ea3e9400890133e57c33f1b3cff646b22)
Oct 14 09:33:53 H systemd-machined[652]:     #1 0x78b2d3e8a659 in image_new /systemd-meson-build/../build/src/shared/discover-image.c:213:19
Oct 14 09:33:53 H systemd-machined[652]:     #2 0x78b2d3e7e771 in image_make /systemd-meson-build/../build/src/shared/discover-image.c:429:21
Oct 14 09:33:53 H systemd-machined[652]:     #3 0x78b2d3e7d52b in image_find /systemd-meson-build/../build/src/shared/discover-image.c:708:29
Oct 14 09:33:53 H systemd-machined[652]:     #4 0x585b1a08bed9 in vl_method_list_images /systemd-meson-build/../build/src/machine/machined-varlink.c:589:21
Oct 14 09:33:53 H systemd-machined[652]:     #5 0x78b2d44da37b in varlink_dispatch_method /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1383:29
Oct 14 09:33:53 H systemd-machined[652]:     #6 0x78b2d44da37b in sd_varlink_process /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1456:13
Oct 14 09:33:53 H systemd-machined[652]:     #7 0x78b2d44f41f5 in defer_callback /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:2880:16
Oct 14 09:33:53 H systemd-machined[652]:     #8 0x78b2d4367bcf in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c
Oct 14 09:33:53 H systemd-machined[652]:     #9 0x78b2d43660d9 in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4849:21
Oct 14 09:33:53 H systemd-machined[652]:     #10 0x78b2d436a039 in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4910:21
Oct 14 09:33:53 H systemd-machined[652]:     #11 0x78b2d3e07515 in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:147:21
Oct 14 09:33:53 H systemd-machined[652]:     #12 0x585b1a06cc1e in run /systemd-meson-build/../build/src/machine/machined.c:346:13
Oct 14 09:33:53 H systemd-machined[652]:     #13 0x585b1a06cc1e in main /systemd-meson-build/../build/src/machine/machined.c:358:1
Oct 14 09:33:53 H systemd-machined[652]:     #14 0x78b2d3634e07  (/usr/lib/libc.so.6+0x25e07) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #15 0x78b2d3634ecb in __libc_start_main (/usr/lib/libc.so.6+0x25ecb) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #16 0x585b1a06a294 in _start (/usr/lib/systemd/systemd-machined+0x34294) (BuildId: 7e799bc6af64f913587709a1e143113c9443cdb3)
Oct 14 09:33:53 H systemd-machined[652]: Indirect leak of 17 byte(s) in 3 object(s) allocated from:
Oct 14 09:33:53 H systemd-machined[652]:     #0 0x78b2d4d52682 in strdup (/usr/lib/clang/18/lib/linux/libclang_rt.asan-x86_64.so+0x152682) (BuildId: 0f6d876ea3e9400890133e57c33f1b3cff646b22)
Oct 14 09:33:53 H systemd-machined[652]:     #1 0x78b2d421ac7c in strv_extend_with_size /systemd-meson-build/../build/src/basic/strv.c:704:13
Oct 14 09:33:53 H systemd-machined[652]:     #2 0x78b2d4136802 in strv_extend /systemd-meson-build/../build/src/basic/strv.h:62:16
Oct 14 09:33:53 H systemd-machined[652]:     #3 0x78b2d4136802 in load_env_file_push_pairs /systemd-meson-build/../build/src/basic/env-file.c:483:13
Oct 14 09:33:53 H systemd-machined[652]:     #4 0x78b2d41340d5 in parse_env_file_internal /systemd-meson-build/../build/src/basic/env-file.c:153:37
Oct 14 09:33:53 H systemd-machined[652]:     #5 0x78b2d413643c in load_env_file_pairs /systemd-meson-build/../build/src/basic/env-file.c:500:13
Oct 14 09:33:53 H systemd-machined[652]:     #6 0x78b2d4136ab4 in load_env_file_pairs_fd /systemd-meson-build/../build/src/basic/env-file.c:518:16
Oct 14 09:33:53 H systemd-machined[652]:     #7 0x78b2d41b7383 in load_extension_release_pairs /systemd-meson-build/../build/src/basic/os-util.c:434:16
Oct 14 09:33:53 H systemd-machined[652]:     #8 0x78b2d3e881ab in load_os_release_pairs /systemd-meson-build/../build/src/basic/os-util.h:54:16
Oct 14 09:33:53 H systemd-machined[652]:     #9 0x78b2d3e881ab in image_read_metadata /systemd-meson-build/../build/src/shared/discover-image.c:1502:21
Oct 14 09:33:53 H systemd-machined[652]:     #10 0x585b1a08d236 in list_image_one_and_maybe_read_metadata /systemd-meson-build/../build/src/machine/machined-varlink.c:520:21
Oct 14 09:33:53 H systemd-machined[652]:     #11 0x585b1a08c082 in vl_method_list_images /systemd-meson-build/../build/src/machine/machined-varlink.c:595:24
Oct 14 09:33:53 H systemd-machined[652]:     #12 0x78b2d44da37b in varlink_dispatch_method /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1383:29
Oct 14 09:33:53 H systemd-machined[652]:     #13 0x78b2d44da37b in sd_varlink_process /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:1456:13
Oct 14 09:33:53 H systemd-machined[652]:     #14 0x78b2d44f41f5 in defer_callback /systemd-meson-build/../build/src/libsystemd/sd-varlink/sd-varlink.c:2880:16
Oct 14 09:33:53 H systemd-machined[652]:     #15 0x78b2d4367bcf in source_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c
Oct 14 09:33:53 H systemd-machined[652]:     #16 0x78b2d43660d9 in sd_event_dispatch /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4849:21
Oct 14 09:33:53 H systemd-machined[652]:     #17 0x78b2d436a039 in sd_event_run /systemd-meson-build/../build/src/libsystemd/sd-event/sd-event.c:4910:21
Oct 14 09:33:53 H systemd-machined[652]:     #18 0x78b2d3e07515 in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:147:21
Oct 14 09:33:53 H systemd-machined[652]:     #19 0x585b1a06cc1e in run /systemd-meson-build/../build/src/machine/machined.c:346:13
Oct 14 09:33:53 H systemd-machined[652]:     #20 0x585b1a06cc1e in main /systemd-meson-build/../build/src/machine/machined.c:358:1
Oct 14 09:33:53 H systemd-machined[652]:     #21 0x78b2d3634e07  (/usr/lib/libc.so.6+0x25e07) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #22 0x78b2d3634ecb in __libc_start_main (/usr/lib/libc.so.6+0x25ecb) (BuildId: 98b3d8e0b8c534c769cb871c438b4f8f3a8e4bf3)
Oct 14 09:33:53 H systemd-machined[652]:     #23 0x585b1a06a294 in _start (/usr/lib/systemd/systemd-machined+0x34294) (BuildId: 7e799bc6af64f913587709a1e143113c9443cdb3)
Oct 14 09:33:53 H systemd-machined[652]: SUMMARY: AddressSanitizer: 2388 byte(s) leaked in 41 allocation(s).
```

ikruglov commented:
> Oct 14 09:33:53 H systemd-machined[652]:     #6 0x78b2d3e881ab in load_os_release_pairs /systemd-meson-build/../build/src/basic/os-util.h:54:16

I'm looking into it

yuwata commented:
-> #34679.
